core/extensions.rs

```diff
@@ -21,6 +21,13 @@ pub enum ExtensionFileSourceCode {
   // embedder is creating snapshots. Files will be loaded from the filesystem
   // during the build time and they will only be present in the V8 snapshot.
   LoadedFromFsDuringSnapshot(PathBuf),
+
+  /// Source code has already been evaluated via the snapshot, and should not be
+  /// loaded. This variant is used to avoid storing snapshot-evaluated code in
+  /// the CLI binary -- however it would most likely need to be removed once we
+  /// support `ShadowRealm`s in which case the extension code always needs to be
+  /// available for bootstrapping other realms.
+  AlreadyEvaluated,
```

Comment:
`ExtensionFileSourceCode::AlreadyEvaluated`. Previously this was represented by just not telling core about esm files that are already evaluated in the passed snapshot, but we need access to the total list of ext specifier names so now it's represented by this variant.

I think supporting `ShadowRealm` will require all extension code at runtime though, in addition to the startup snapshot, which would incur some binary cost.

Reply:
Why do you think we need to require all extension code for `ShadowRealm`? Wouldn't we start up from the same snapshot there? 

Snapshots can store multiple different contexts with different globals variables - we can leverage this feature to selectively make some extensions available.

Reply:
> Snapshots can store multiple different contexts with different globals variables

My understanding of this is that you can have bootstrapping code like `globalThis.friendRealm = new ShadowRealm()`, snapshot that, load the snapshot and still have both realms. But there's no way of using an isolate-wide startup snapshot to start a new realm in that same isolate, is there?

Reply:
Snapshots seem to be of a context, not of an isolate as a whole somehow; and in my testing, it seemed like if there is a snapshot for the isolate, that will be applied by default when a context is created.

After all, V8 uses snapshots internally for initializing the JS built-ins (though written in CSA and Torque, not JS), and creating a context creates different built-ins – in terms of object identity, at least.

Reply:
Andreu is right here - currently we use the first approach (ie. the snapshot of a isolate), but we have all the necessary APIs to have snapshots of different contexts. @crowlKats and I planned to experiment with the second approach so I think we shouldn't discard it yet. But definitely something that can be addressed in a follow up.

Reply:
Good to know, I think that means my `This should be per-realm` TODOs aren't applicable